### PR TITLE
Clean up evaluation algo code

### PR DESCRIPTION
**Pull request by lizgzil**
# Description
- Changes to evaluate algorithm files to account for new fuzzymatcher which was in master
- Add number in sample to outputs
- Make output txt file look at bit nicer - this involves restructuring the existing calculation results
- change names of evaluations to be descriptive (rather than 1,2,..)
- logger logs how long each evaluation takes (approximately 5 mins in total)
- changing names to find, split, parse, match nomenclature 
- merging tests 1 and 2

# Output
```
Date of evaluation = 2019-07-19-1522

-----How well the scraper finds the references section:-----

Score
0.388

Lenient F1-score (references section exists or not)
0.944

Metrics by provider
Provider                                         msf    nice  unicef  who_iris        all
Number of pdfs included                       15.000  17.000  31.000    32.000  95.000000
Number of pdfs with sections text              7.000   5.000   9.000     5.000  26.000000
Proportion of pdfs with sections text          0.467   0.294   0.290     0.156   0.274000
Lenient F1 score for all sections included     0.911   0.941   0.914     0.990   0.944000
Number of pdfs with a bibliograph section      1.000   0.000   0.000     1.000   2.000000
Number of pdfs with a endnotes section         2.000   0.000   4.000     0.000   6.000000
Number of pdfs with a reference section        4.000   5.000   5.000     4.000  18.000000
Proportion with a bibliograph section          0.067   0.000   0.000     0.031        NaN
Proportion with a endnotes section             0.133   0.000   0.129     0.000        NaN
Proportion with a reference section            0.267   0.294   0.161     0.125        NaN
Lenient F1 score for the bibliograph section   0.933   1.000   1.000     1.000   0.989474
Lenient F1 score for the endnotes section      0.933   1.000   1.000     1.000   0.989474
Lenient F1 score for the reference section     0.867   0.824   0.742     0.969   0.852632

Number of pdfs with sections text
26

Mean normalised Levenshtein distance
0.612

Strict accuracy (micro)
0.0

Lenient accuracy (micro) (normalised Levenshtein < 0.3)
0.308

Mean normalised Levenshtein distance for the bibliograph section
0.506

Strict accuracy for the bibliograph section
0.0

Lenient accuracy (normalised Levenshtein< 0.3) for the bibliograph section
0.5

Mean normalised Levenshtein distance for the reference section
0.652

Strict accuracy for the reference section
0.0

Lenient accuracy (normalised Levenshtein< 0.3) for the reference section
0.278

Mean normalised Levenshtein distance for the endnotes section
0.524

Strict accuracy for the endnotes section
0.0

Lenient accuracy (normalised Levenshtein< 0.3) for the endnotes section
0.333

-----How well the splitter predicted how many references there were:-----

Score
0.614

Number of reference sections in sample
70

Proportion below threshold of 40
0.614

Median difference
20.5

Metrics grouped by source
Source                                    msf  nice  unicef  who_iris
Number of reference sections in sample  9.000  25.0  14.000    22.000
Median difference                       7.300  95.7  27.400     8.700
Proportion below threshold of 40        0.889   0.2   0.786     0.864

-----How well the parser predicted reference component texts:-----

Score
0.862

Number of references in sample
205

Number of non-blank reference components
935

Strict accuracy (micro)
0.403

Lenient accuracy (micro)(normalised Levenshtein < 0.3)
0.817

Mean normalised Levenshtein distance (all categories)
0.138

Metrics per category
                                                Authors   Issue  Journal  Pagination  PubYear    Title  Volume
Number of non-blank reference components        176.000  40.000  154.000      90.000  188.000  202.000  85.000
Strict accuracy                                   0.341   0.639    0.337       0.546    0.293    0.244   0.420
Lenient accuracy(normalised Levenshtein < 0.3)    0.615   0.917    0.839       0.912    0.956    0.610   0.868
Mean normalised Levenshtein distance              0.299   0.074    0.137       0.055    0.036    0.256   0.108

-----How well the matcher matched references:-----

Score
0.942

Number of references in sample
2000

Micro average F1-score
0.942

Micro average recall
0.942

Micro average precision
0.942

Classification report
              precision    recall  f1-score   support

       False       0.90      1.00      0.94      1000
        True       1.00      0.89      0.94      1000

   micro avg       0.94      0.94      0.94      2000
   macro avg       0.95      0.94      0.94      2000
weighted avg       0.95      0.94      0.94      2000


Frequency table of match types
What did we expect?    Match expected  No match expected
What did we find?                                       
Correct match found               886                  0
Incorrect match found               0                  3
No match found                    114                997
```
# How Has This Been Tested?

Not tested.

### CODE REVIEW

File: policytool/refparse/algo_evaluation/evaluate_match_references.py [hunk 1]
```diff
@@ -22,17 +22,20 @@ def predict_match_data(matcher, match_data):
...
                 column gives the score of how similar the matched 
                 reference titles are.
     """
 
     predictions = []
     for i, ref in match_data.iterrows():
-        matched_publications = matcher.match_vectorised(ref)
-        predictions.append(matched_publications['uber_id'][0])
+        matched_publications = matcher.match(ref)
+        if matched_publications:
```
Comment (nsorros) [inline policytool/refparse/algo_evaluation/evaluate_match_references.py]: is that needed? why does it return when this fails?

File: policytool/refparse/algo_evaluation/evaluate_match_references.py [hunk 2]
```diff
@@ -93,7 +96,7 @@ def evaluate_metric(actual, predicted):
     return metrics
 
 def evaluate_match_references(
-        evaluation_references, match_threshold, sample_N
+        evaluation_references, match_threshold, length_threshold, sample_N
```
Comment (nsorros) [inline policytool/refparse/algo_evaluation/evaluate_match_references.py]: so you solve the problem by reintroducing a parameter that is not used?

File: policytool/refparse/utils/fuzzy_match.py [hunk 3]
```diff
@@ -32,7 +32,7 @@ def search_publications(self, reference, nb_results=10):
         return retrieved_publications[:nb_results]
 
     def match(self, reference):
-        if not reference:
+        if reference.empty:
```
Comment (nsorros) [inline policytool/refparse/utils/fuzzy_match.py]: why? this assumes a dataframe?

File: policytool/refparse/algo_evaluation/evaluate_parse.py [hunk 4]
```diff
@@ -43,16 +43,24 @@ def evaluate_metric(
     accuracy_quite_equal_cat = quite_equal.mean()
     accuracy_quite_equal = quite_equal.values.mean()
 
+    number_sample = (actual_categories!="").sum()
```
Comment (nsorros) [inline policytool/refparse/algo_evaluation/evaluate_parse.py]: sample_number_per_category is a better name if this is not a single number

**Comment by lizgzil** [comment]
@ivyleavedtoadflax Strange, it's definitely not supposed to be duplicated! 
> * policytool/algo_evaluation/data_evaluate
I don't find this locally or on master. Where did you see this?

File: policytool/refparse/algo_evaluation/evaluate_match_references.py (hunk 1 above)
Comment (lizgzil) [reply policytool/refparse/algo_evaluation/evaluate_match_references.py]: the new matcher doesn't return anything if there is no match (in the past it returned a dict of None)

File: policytool/refparse/algo_evaluation/evaluate_match_references.py (hunk 2 above)
Comment (lizgzil) [reply policytool/refparse/algo_evaluation/evaluate_match_references.py]: it is needed for the new fuzzymatcher, unless you think I should just use the default length threshold?

File: policytool/refparse/utils/fuzzy_match.py (hunk 3 above)
Comment (lizgzil) [reply policytool/refparse/utils/fuzzy_match.py]: yep

**Comment by ivyleavedtoadflax** [comment]
> @ivyleavedtoadflax Strange, it's definitely not supposed to be duplicated!
> 
> > * policytool/algo_evaluation/data_evaluate
> >   I don't find this locally or on master. Where did you see this?

Ah sorry my bad it was just on my local repo. Not sure how that happened. I assume it was made by a script.

File: policytool/refparse/algo_evaluation/evaluation.md [hunk 5]
```diff
@@ -11,23 +11,24 @@ In this document we describe how we got each of the evaluation datasets uses in
 - The text was copied and pasted from the pdf (starting from the section name to the last character of the section) into a text editor and saved as a .txt file in the relevant sections folder (e.g. "./algo_evaluation/data_evaluate/pdf_sections/reference") and saved with the same unique name as the pdf was.
 
 
-### Evaluation Scores 1 and 2
+### Evaluation Scores
```
Comment (nsorros) [inline policytool/refparse/algo_evaluation/evaluation.md]: Find evaluation

File: policytool/refparse/algo_evaluation/evaluation.md [hunk 6]
```diff
@@ -11,23 +11,24 @@ In this document we describe how we got each of the evaluation datasets uses in
...
 ### Data
 - Randomly sample policy documents with scraped references sections.
 - Made a decision about the quality of the scrape - if it was bad then we wouldn't include, but if the references text scraped included the whole references text we took it forward. We have 6 documents from MSF documents, 20 from NICE, 10 from UNICEF, and 17 from WHO documents.
 - Make a decision about where the references started and ended. There was often a lot of non-references text scraped after the references.
 - Copy and paste the references section only into a txt file (with the document hash as the file name) stored in the "scraped_references_sections" folder.
 - Manually count the number of references from this document's pdf, record in "split_section_test_data.csv".
 
-
-### Evaluation Score 3
+### Evaluation Scores
```
Comment (nsorros) [inline policytool/refparse/algo_evaluation/evaluation.md]: Parse evaluation

**Comment by nsorros** [comment]
I am still getting `No such file or directory: '/Users/sorrosn/code/policytool/policytool/refparse/algo_evaluation/../algo_evaluation/data_evaluate/pdf_providers.csv'`

**Comment by lizgzil** [comment]
> I am still getting `No such file or directory: '/Users/sorrosn/code/policytool/policytool/refparse/algo_evaluation/../algo_evaluation/data_evaluate/pdf_providers.csv'`

@aCampello did you get any issues like this?

**Comment by aCampello** [comment]
Yes. That is because of a typo in the `pdf_providers` name. I just pushed a quick fix for this.

**Comment by hblanks** [comment]
@lizgzil  - it seems this might have broken the build.

**Comment by nsorros** [comment]
@lizgzil broken build becomes priority number one always as engineers and anyone pushing code will trigger broken builds.

**Pull request merged**